Custom Error type returned from helper function is not assignable to type error

The following program `sample.go` triggers a panic:

```go
package main

import "fmt"

type CustomError string

func (s CustomError) Error() string {
        return string(s)
}

func NewCustomError(errorText string) CustomError {
        return CustomError(errorText)
}

func fail() (err error) {
        return NewCustomError("Everything is going wrong!")
}

func main() {
        fmt.Println(fail())
}
```

Expected result:
```console
$ go run ./sample.go
Everything is going wrong!
```

Got:
```console
$ yaegi ./sample.go
sample.go:12:2: panic
sample.go:20:14: panic
sample.go:20:14: panic
run: reflect.Set: value of type string is not assignable to type error
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0xc0002d6a00, 0x40)
        /usr/local/go/src/runtime/debug/stack.go:24 +0x9f
github.com/traefik/yaegi/interp.(*Interpreter).eval.func1(0xc0001cbc68)
        /go/src/yaegi/interp/interp.go:494 +0xc8
panic(0xe1e0c0, 0xc00006cc90)
        /usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/traefik/yaegi/interp.runCfg.func1(0xc00035a160, 0xc000464000, 0xc0001cba48)
        /go/src/yaegi/interp/run.go:185 +0x253
panic(0xe1e0c0, 0xc00006cc90)
        /usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/traefik/yaegi/interp.runCfg.func1(0xc00035a210, 0xc000463500, 0xc0001cb838)
        /go/src/yaegi/interp/run.go:185 +0x253
panic(0xe1e0c0, 0xc00006cc90)
        /usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/traefik/yaegi/interp.runCfg.func1(0xc00035a2c0, 0xc000462700, 0xc0001cb628)
        /go/src/yaegi/interp/run.go:185 +0x253
panic(0xe1e0c0, 0xc00006cc90)
        /usr/local/go/src/runtime/panic.go:965 +0x1b9
reflect.Value.assignTo(0xe1e0c0, 0xc00006cc70, 0x198, 0xf5d27a, 0xb, 0xe71520, 0xc00006cc40, 0xc00006cc80, 0x98, 0xc00006cc80)
        /usr/local/go/src/reflect/value.go:2451 +0x3f7
reflect.Value.Set(0xe71520, 0xc00006cc40, 0x194, 0xe1e0c0, 0xc00006cc70, 0x198)
        /usr/local/go/src/reflect/value.go:1564 +0xbd
github.com/traefik/yaegi/interp._return.func1(0xc00035a2c0, 0xc00006cbd0)
        /go/src/yaegi/interp/run.go:2272 +0x85
github.com/traefik/yaegi/interp.runCfg(0xc000462700, 0xc00035a2c0)
        /go/src/yaegi/interp/run.go:191 +0x87
github.com/traefik/yaegi/interp.call.func7(0xc00035a210, 0x15aa880)
        /go/src/yaegi/interp/run.go:1298 +0xd05
github.com/traefik/yaegi/interp.runCfg(0xc000463500, 0xc00035a210)
        /go/src/yaegi/interp/run.go:191 +0x87
github.com/traefik/yaegi/interp.call.func7(0xc00035a160, 0xc00035d5e0)
        /go/src/yaegi/interp/run.go:1298 +0xd05
github.com/traefik/yaegi/interp.runCfg(0xc000464000, 0xc00035a160)
        /go/src/yaegi/interp/run.go:191 +0x87
github.com/traefik/yaegi/interp.(*Interpreter).run(0xc000358000, 0xc000463800, 0xc00035a000)
        /go/src/yaegi/interp/run.go:122 +0x2b0
github.com/traefik/yaegi/interp.(*Interpreter).eval(0xc000358000, 0xc00045c140, 0x13d, 0x7ffd7afa3eff, 0x9, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/yaegi/interp/interp.go:586 +0x80c
github.com/traefik/yaegi/interp.(*Interpreter).EvalPath(0xc000358000, 0x7ffd7afa3eff, 0x9, 0xc00045c000, 0x13d, 0x0, 0x0, 0xc000095040)
        /go/src/yaegi/interp/interp.go:408 +0x105
main.runFile(0xc000358000, 0x7ffd7afa3eff, 0x9, 0x1, 0x1)
        /go/src/yaegi/cmd/yaegi/run.go:123 +0xae
main.run(0xc000140010, 0x1, 0x1, 0xc00010c101, 0xc000102480)
        /go/src/yaegi/cmd/yaegi/run.go:89 +0x86e
main.main()
        /go/src/yaegi/cmd/yaegi/yaegi.go:144 +0x415
```

Made sample based on this real world code: https://github.com/dgrijalva/jwt-go/blob/e58d3b75485401073a004dd9858860f8a5d9f627/errors.go#L31

--- a/_test/issue-1091.go
+++ b/_test/issue-1091.go
@@ -1,0 +1,24 @@
+package main
+
+import "fmt"
+
+type CustomError string
+
+func (s CustomError) Error() string {
+	return string(s)
+}
+
+func NewCustomError(errorText string) CustomError {
+	return CustomError(errorText)
+}
+
+func fail() (err error) {
+	return NewCustomError("Everything is going wrong!")
+}
+
+func main() {
+	fmt.Println(fail())
+}
+
+// Output:
+// Everything is going wrong!

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -1004,7 +1004,9 @@ func (interp *Interpreter) cfg(root *node, importPath string) ([]*node, error) {
 				}
 				if typ := n.child[0].typ; len(typ.ret) > 0 {
 					n.typ = typ.ret[0]
-					if n.anc.kind == returnStmt {
+					if n.anc.kind == returnStmt && n.typ.id() == sc.def.typ.ret[0].id() {
+						// Store the result to directly to the return value area of frame.
+						// It can be done only if no type conversion at return is involved.
 						n.findex = childPos(n)
 					} else {
 						n.findex = sc.add(n.typ)

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -1005,7 +1005,7 @@ func (interp *Interpreter) cfg(root *node, importPath string) ([]*node, error) {
 				if typ := n.child[0].typ; len(typ.ret) > 0 {
 					n.typ = typ.ret[0]
 					if n.anc.kind == returnStmt && n.typ.id() == sc.def.typ.ret[0].id() {
-						// Store the result to directly to the return value area of frame.
+						// Store the result directly to the return value area of frame.
 						// It can be done only if no type conversion at return is involved.
 						n.findex = childPos(n)
 					} else {

--- a/interp/run.go
+++ b/interp/run.go
@@ -2259,12 +2259,16 @@ func _return(n *node) {
 	case 0:
 		n.exec = nil
 	case 1:
-		// This is an optimisation that is applied for binary expressions or function
-		// calls, but not for (binary) expressions involving const, as the values are not
-		// stored in the frame in that case.
-		if !child[0].rval.IsValid() && child[0].kind == binaryExpr || isCall(child[0]) {
+		switch {
+		case !child[0].rval.IsValid() && child[0].kind == binaryExpr:
+			// No additional runtime operation is necessary for constants (not in frame) or
+			// binary expressions (stored directly at the right location in frame).
 			n.exec = nil
-		} else {
+		case isCall(child[0]) && n.child[0].typ.id() == def.typ.ret[0].id():
+			// Calls are optmized as long as no type conversion is involved.
+			n.exec = nil
+		default:
+			// Regular return: store the value to return at to start of the frame.
 			v := values[0]
 			n.exec = func(f *frame) bltn {
 				f.data[0].Set(v(f))

